packages/flutter_tools/lib/src/commands/symbolize.dart

```diff
@@ -0,0 +1,153 @@
+
+  @override
+  String get name => 'symbolize';
+
+  @override
+  bool get shouldUpdateCache => false;
+
+  @override
+  Future<void> validateCommand() {
+    if (!argResults.wasParsed('debug-info')) {
+      throwToolExit('"--debug-info" is required to symbolicate stack traces.');
+    }
+    if (!_fileSystem.isFileSync(stringArg('debug-info'))) {
+      throwToolExit('${stringArg('debug-info')} does not exist.');
+    }
+    if (argResults.wasParsed('input') && !_fileSystem.isFileSync(stringArg('input'))) {
+      throwToolExit('${stringArg('input')} does not exist.');
+    }
+    return super.validateCommand();
+  }
+
+  @override
+  Future<FlutterCommandResult> runCommand() async {
+    Stream<List<int>> input;
+    IOSink output;
+
+    // Configure output to either specified file or stdout.
+    if (argResults.wasParsed('output')) {
+      final File outputFile = _fileSystem.file(stringArg('output'));
+      if (!outputFile.parent.existsSync()) {
+        outputFile.parent.createSync(recursive: true);
+      }
+       output = outputFile.openWrite();
+    } else {
+      final StreamController<List<int>> outputController = StreamController<List<int>>();
+      outputController
+        .stream
+        .transform(utf8.decoder)
+        .listen(_stdio.stdoutWrite);
+      output = IOSink(outputController);
+    }
+
+    // Configure input from either specified file or stdin.
+    if (argResults.wasParsed('input')) {
+      input = _fileSystem.file(stringArg('input')).openRead();
+    } else {
+      input = _stdio.stdin;
+    }
+
+    final Uint8List symbols = _fileSystem.file(stringArg('debug-info')).readAsBytesSync();
+    await _dwarfSymbolicationService.decode(
+      input: input,
+      output: output,
+      symbols: symbols,
+    );
+
+    return FlutterCommandResult.success();
+  }
+}
+
+/// A service which decodes stack traces from Dart applications.
+class DwarfSymbolicationService {
+  const DwarfSymbolicationService();
+
+  /// Decode a stack trace from [input] and place the results in [output].
+  ///
+  /// Requires [symbols] to be a buffer created from the `--split-debug-info`
+  /// command line flag.
+  ///
+  /// Throws a [ToolExit] if the symbols cannot be parsed or the stack trace
+  /// cannot be decoded.
+  Future<void> decode({
+    @required Stream<List<int>> input,
+    @required IOSink output,
+    @required Uint8List symbols,
+  }) async {
+    final Dwarf dwarf = Dwarf.fromElf(Elf.fromReader(Reader.fromTypedData(symbols)));
+    if (dwarf == null) {
+      throwToolExit('Failed to decode symbols file');
+    }
+
+    final Completer<void> onDone = Completer<void>();
+    input
+      .transform(const Utf8Decoder())
+      .transform(const LineSplitter())
+      .transform(DwarfStackTraceDecoder(dwarf, includeInternalFrames: true))
+      .listen(output.writeln, onDone: onDone.complete, onError: onDone.completeError);
```

Comment:
If the `writeln` throws, it'll be thrown into the Zone. You can handle exceptions from `writeln` like:
```dart
StreamSubscription sub;
sub = input
  .transform(const Utf8Decoder())
  .transform(const LineSplitter())
  .transform(DwarfStackTraceDecoder(dwarf, includeInternalFrames: true))
  .listen((thing) async {
    try {
      output.writeln(thing);
    } on Exception catch(e, s) {
      sub.cancel().whenComplete(() {
        if (!onDone.isCompleted) {
          onDone.completeError(e, s);
        }
      );
    }
  }, onDone: onDone.complete, onError: onDone.completeError);
```